src/main/scala/io/digitallibrary/language/model/LanguageTag.scala

```diff
@@ -0,0 +1,70 @@
+/*
+ * Part of GDL language.
+ * Copyright (C) 2017 Global Digital Library
+ *
+ * See LICENSE
+ */
+
+package io.digitallibrary.language.model
+
+import io.digitallibrary.language.model.CodeLists.{Iso15924, Iso3166, Iso639}
+
+import scala.util.{Failure, Try}
+
+case class LanguageTag (language: Iso639, script: Option[Iso15924], region: Option[Iso3166]) {
+
+  override def toString: String = {
+    Seq(Some(language.id), script.map(_.code), region.map(_.code)).flatten.mkString("-").toLowerCase
+
+  }
+
+  def displayName: String = {
+    val scriptAndRegion = (script.map(_.englishName) :: region.map(_.name) :: Nil).flatten.mkString(", ")
+    if (scriptAndRegion.isEmpty) {
+      language.refName
+    } else {
+      s"${language.refName} ($scriptAndRegion)"
+    }
+  }
+}
+
+object LanguageTag {
+  def apply(languageTagAsString: String): LanguageTag = {
+    val parts = languageTagAsString.split("-")
+    val tag = parts.size match {
+      case 1 => withLanguage(parts(0))
+      case 2 if parts(1).length == 2 => withLanguageAndRegion(parts(0), parts(1))
+      case 2 if parts(1).length == 4 => withLanguageAndScript(parts(0), parts(1))
+      case 3 => withLanguageScriptAndRegion(parts(0), parts(1), parts(2))
+      case _ => Failure(new LanguageNotSupportedException(s"The language tag '$languageTagAsString' is not supported."))
+    }
```

Comment:
Instead of using indices, would it read better like this?
```scala
    val tag = languageTagAsString.split("-").toList match {
      case lang :: Nil => withLanguage(lang)
      case lang :: regionOrScript :: Nil if regionOrScript.length == 2 => withLanguageAndRegion(lang, regionOrScript)
      case lang :: regionOrScript :: Nil if regionOrScript.length == 4 => withLanguageAndScript(lang, regionOrScript)
      case lang :: script :: region :: Nil => withLanguageScriptAndRegion(lang, script, region)
      case _ => Failure(new LanguageNotSupportedException(s"The language tag '$languageTagAsString' is not supported."))
    }
```

Reply:
Yes, only modification to your suggestion is to rename `regionOrScript` to `region` in the case where region is expected, and `script` where script is expected.

👍